src/hash_map.c

```diff
@@ -0,0 +1,562 @@
+      return false;
+    }
+    // Check that the hashes match first as that will be the quicker comparison to quick fail on
+    if (bucket_entry->hashed_key == *key_hash &&
+      (0 == hash_map->impl->key_cmp_func(bucket_entry->key, key)))
+    {
+      *bucket_index = i;
+      *entry = bucket_entry;
+      return true;
+    }
+  }
+
+  return false;
+}
+
+rcutils_ret_t
+rcutils_hash_map_set(rcutils_hash_map_t * hash_map, const void * key, const void * value)
+{
+  HASH_MAP_VALIDATE_HASH_MAP(hash_map);
+  HASH_MAP_VALIDATE_ARGUMENT_NOT_NULL(key);
+  HASH_MAP_VALIDATE_ARGUMENT_NOT_NULL(value);
+
+  size_t key_hash = 0, map_index = 0, bucket_index = 0;
+  bool already_exists = false;
+  rcutils_hash_map_entry_t * entry = NULL;
+  rcutils_ret_t ret = RCUTILS_RET_OK;
+
+  already_exists = hash_map_find(hash_map, key, &key_hash, &map_index, &bucket_index, &entry);
+
+  if (already_exists) {
+    // Just update the existing value to match the new value
+    memcpy(entry->value, value, hash_map->impl->data_size);
+  } else {
+    // We need to create a new entry in the map
+    rcutils_allocator_t * allocator = &hash_map->impl->allocator;
+
+    // Start by trying to allocate the memory we need for the new entry
+    entry = allocator->allocate(sizeof(rcutils_hash_map_entry_t), allocator->state);
+    if (NULL == entry) {
+      return RCUTILS_RET_BAD_ALLOC;
+    }
+    entry->key = allocator->allocate(hash_map->impl->key_size, allocator->state);
+    entry->value = allocator->allocate(hash_map->impl->data_size, allocator->state);
+    if (NULL == entry->key || NULL == entry->value) {
+      ret = RCUTILS_RET_BAD_ALLOC;
+    }
+
+    // Set the entry data and try to insert into the bucket
+    if (RCUTILS_RET_OK == ret) {
+      entry->hashed_key = key_hash;
+      memcpy(entry->value, value, hash_map->impl->data_size);
+      memcpy(entry->key, key, hash_map->impl->key_size);
+
+      bucket_index = key_hash % hash_map->impl->capacity;
+      ret = hash_map_insert_entry(hash_map->impl->map, bucket_index, entry, allocator);
+    }
+
+    if (RCUTILS_RET_OK != ret) {
+      // If somethign went wrong somewhere then cleanup the memory we've allocated
+      hash_map_deallocate_entry(allocator, entry);
+      return ret;
+    } else {
+      hash_map->impl->size++;
+    }
+  }
+
+  // Time to check if we've exceeded our Load Factor and grow the map if so
+  ret = hash_map_check_and_grow_map(hash_map);
+  // Just log on this failure because the map can continue to operate with degraded performance
+  RCUTILS_LOG_ERROR_EXPRESSION(RCUTILS_RET_OK != ret, "Failed to grow hash_map. Reason: %d", ret);
+
+  return RCUTILS_RET_OK;
+}
+
+rcutils_ret_t
+rcutils_hash_map_unset(rcutils_hash_map_t * hash_map, const void * key)
+{
+  HASH_MAP_VALIDATE_HASH_MAP(hash_map);
+  HASH_MAP_VALIDATE_ARGUMENT_NOT_NULL(key);
+
+  size_t key_hash = 0, map_index = 0, bucket_index = 0;
+  bool already_exists = false;
```

Comment:
@nburek nit: why not declaring it at the time of first use?

Reply:
Same as above.